### PR TITLE
New: rule indent

### PR DESCRIPTION
**Pull request by valorkin**
Added new rule: indent
merge of https://github.com/jscs-dev/node-jscs/blob/master/lib/rules/validate-indentation.js
to eslint rule

in order to solve #1022

### CODE REVIEW

**Comment by es128** [comment]
haha, why me?

File: docs/rules/README.md [hunk 1]
```diff
@@ -124,12 +124,13 @@ These rules are specific to JavaScript running on Node.js.
...
 These rules are purely matters of style and are quite subjective.
 
+* [indent](indent.md) - this option sets a specific tab width for your code (off by default)
 * [brace-style](brace-style.md) - enforce one true brace style (off by default)
 * [camelcase](camelcase.md) - require camel case names
 * [comma-spacing](comma-spacing.md) - enforce spacing before and after comma
 * [comma-style](comma-style.md) - enforce one true comma style (off by default)
 * [consistent-this](consistent-this.md) - enforces consistent naming when capturing the current execution context (off by default)
-* [eol-last](eol-last.md) - enforce newline at the end of file, with no multiple empty lines
+* [eol-last](eol-last.md) - enforce newline acd t the end of file, with no multiple empty lines
```
Comment (nzakas) [inline docs/rules/README.md]: Mistake?

File: lib/rules/indent.js [hunk 2]
```diff
@@ -0,0 +1,378 @@
+/*eslint no-use-before-define:[2, "nofunc"]*/
+/**
+ * @fileoverview This option sets a specific tab width for your code
+ * @author Dmitriy Shekhovtsov
+ * @copyright 2014 Dmitriy Shekhovtsov. All rights reserved.
```
Comment (nzakas) [inline lib/rules/indent.js]: We need to be sure to credit JSCS in this file.

File: lib/rules/indent.js [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+        SwitchCase: "consequent"
+    };
+    var breakIndents = null;
+
+    //--------------------------------------------------------------------------
+    // Helpers
+    //--------------------------------------------------------------------------
+
+    // any helper functions should go here or else delete this section
+    function isMultiline(node) {
```
Comment (nzakas) [inline lib/rules/indent.js]: We require JSDoc comments for all functions.

File: tests/lib/rules/indent.js [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+            "    default:\n" +
+            "        a();\n" +
+            "        break;\n" +
+            "}",
+            args: [2, 4],
+            errors: expectedErrors([10, 4])
+        },
+        {
+            code:
+            "switch(value){\n" +
```
Comment (nzakas) [inline tests/lib/rules/indent.js]: I think we need an option of whether or not to indent `SwitchCase`. I prefer to indent them, other don't....we need to support both.

File: tests/lib/rules/indent.js [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+                [454, 2],
+                [455, 2],
+                [461, 6],
+                [462, 6],
+                [467, 6],
+                [472, 6],
+                [486, 2],
+                [488, 2],
+                [534, 6],
+                [541, 6]
```
Comment (lo1tuma) [inline tests/lib/rules/indent.js]: Lot of magic numbers :confused:

**Comment by nzakas** [comment]
Wow, this looks like a really good start. Please be sure to take a moment and read over our contributor guidelines: http://eslint.org/docs/developer-guide/contributing.html

File: lib/rules/indent.js [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+        "IfStatement",
+        "WhileStatement",
+        "DoWhileStatement",
+        "ForStatement",
+        "ForInStatement",
+        "ForOfStatement",
+        "FunctionDeclaration",
+        "FunctionExpression",
+        "ArrowExpression",
+        "CatchClause"
```
Comment (lo1tuma) [inline lib/rules/indent.js]: What about `WithStatement`? There is also no testcase which uses `with`.

File: docs/rules/README.md (hunk 1 above)
Comment (MoOx) [inline docs/rules/README.md]: looks like one :)

**Comment by MoOx** [comment]
Except some missing comments, this look really good. Looking forward to use that :)

File: tests/lib/rules/indent.js (hunk 5 above)
Comment (valorkin) [inline tests/lib/rules/indent.js]: @lo1tuma expected errors in fixture [`line number`, `expected indentation`]

File: lib/rules/indent.js (hunk 6 above)
Comment (valorkin) [inline lib/rules/indent.js]: @lo1tuma what is `with`? :) I hope it will be deprecated some day,
got your point anyway

**Comment by gyandeeps** [comment]
Hey @valorkin , here some code if you want to cover some more scenarios. I know this has been out there for a long time but never got integrated to `eslint`.
https://github.com/nodeca/eslint-plugin-nodeca/blob/master/lib/indent.js

**Comment by valorkin** [comment]
@gyandeeps thx, I will check it

File: tests/lib/rules/indent.js [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+            "    break;\n" +
+            "    case \"2\":\n" +
+            "        a();\n" +
+            "    break;\n" +
+            "    default:\n" +
+            "        a();\n" +
+            "        break;\n" +
+            "}",
+            args: [2, 4],
+            errors: expectedErrors([10, 4])
```
Comment (lo1tuma) [inline tests/lib/rules/indent.js]: Shouldn’t this be 2 errors? One in line 4 and one in line 7? Shouldn’t the indentation be 8 instead of 10?

Comment (lo1tuma) [inline tests/lib/rules/indent.js]: I think we should not do this strange JSCS behavior at all. The following example is valid in JSCS with 4 spaces identation

```js
switch (a) {
    case "foo":
        a();
    break;
    case "bar":
    a(); break;
}
```

But this fails:
```js
switch (a) {
    case "foo":
        a();
    break;
    case "bar":
        a(); break;
}
```

Comment (nzakas) [inline tests/lib/rules/indent.js]: Agreed. I'd rather be explicit about exceptions like this rather than guessing.

Comment (valorkin) [inline tests/lib/rules/indent.js]: @nzakas @lo1tuma 
I will be working on this tomorrow

**Comment by joestump** [comment]
@valorkin thanks for coding this up; stoked to add it to my projects. :cake: :beers:

File: tests/lib/rules/indent.js (hunk 4 above)
Comment (valorkin) [inline tests/lib/rules/indent.js]: @nzakas added an option: switchCase, disabled by default
covered with tests

File: lib/rules/indent.js (hunk 6 above)
Comment (valorkin) [inline lib/rules/indent.js]: @lo1tuma tests for WithStatement added

File: lib/rules/indent.js [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+    function markCheck(checkNode) {
+        linesToCheck[checkNode.loc.start.line - 1].check = true;
+    }
+
+    /**
+     * Mark child nodes to be checked later of targeted node,
+     * only if child node not in same line as targeted one
+     * (if child and parent nodes wrote in single line)
+     * @param {ASTNode} node - targeted node
+     * @returns {undefined} - no return
```
Comment (nzakas) [inline lib/rules/indent.js]: Should be `@returns {void}` (no description)

File: lib/rules/indent.js [hunk 9]
```diff
@@ -0,0 +1,478 @@
...
+        }
+
+        if (line.popExtra) {
+            indentStack.pop();
+        }
+
+        return expected;
+    }
+
+    function pushExpectedIndentations(line, actualIndentation) {
```
Comment (nzakas) [inline lib/rules/indent.js]: Missing JSDoc here.

File: lib/rules/indent.js [hunk 10]
```diff
@@ -0,0 +1,478 @@
...
+            indentChar = "\t";
+            indentSize = 1;
+        } else if (typeof context.options[0] === "number") {
+            indentSize = context.options[0];
+        }
+
+        if (context.options[1]) {
+            var opts = context.options[1];
+            options.switchCase = opts.switchCase === true;
+            options.withStatement = opts.withStatement === true;
```
Comment (nzakas) [inline lib/rules/indent.js]: `WithStatement` should always be included in `blockParents` below. There's no reason to force people to turn it on.

File: lib/rules/indent.js [hunk 11]
```diff
@@ -0,0 +1,478 @@
...
+        if (context.options[0] === "tab") {
+            indentChar = "\t";
+            indentSize = 1;
+        } else if (typeof context.options[0] === "number") {
+            indentSize = context.options[0];
+        }
+
+        if (context.options[1]) {
+            var opts = context.options[1];
+            options.switchCase = opts.switchCase === true;
```
Comment (nzakas) [inline lib/rules/indent.js]: Maybe should be `indentSwitchCase` to both be clearer and use the proper capitalization for the node type?

File: lib/rules/indent.js [hunk 12]
```diff
@@ -0,0 +1,478 @@
+/*eslint no-use-before-define:[2, "nofunc"]*/
```
Comment (nzakas) [inline lib/rules/indent.js]: This should come after the fileoverview comment.

File: lib/rules/indent.js [hunk 13]
```diff
@@ -0,0 +1,478 @@
...
+/**
+ * @fileoverview This option sets a specific tab width for your code
+ * @author Dmitriy Shekhovtsov
+ * @author Mike Sherov <mikesherov>
+ * @copyright 2015 Dmitriy Shekhovtsov. All rights reserved.
```
Comment (nzakas) [inline lib/rules/indent.js]: Remove `@author` for Mike, and instead add his copyright under yours:

`@copyright 2014 Mike Sherov.`

We also should mention in this fileoverview comment that this file has been ported and modified from JSCS.

**Comment by lo1tuma** [comment]
A few more concerns about the jscs indent rule:

* Multiline expression statements are not supported

```js
if (a) {
    (1 + 2 +
3); // no error on this line
}
```

* No support for multiline variable declarations

```js
function foo() {
    var a, // this line is validated
b; // this is not validated
}
```
* Comments are not supported, which I think should be at least an option

**Comment by lo1tuma** [comment]
I’m not that happy that we are blindly copying the indent logic from jscs. We don’t know what edge cases they support or don’t, so we can’t even document it.

File: tests/lib/rules/indent.js (hunk 4 above)
Comment (lo1tuma) [inline tests/lib/rules/indent.js]: I thought the `switchCase` option should enforce alignment of `case` and `switch`. Looking at the tests it enforces alignment of `break` with `case`.

File: tests/lib/rules/indent.js (hunk 5 above)
Comment (lo1tuma) [inline tests/lib/rules/indent.js]: I think it is not a good idea to have such a large fixture file. If you add a line somewhere in the beginning of the fixture you have do adjust all those magic numbers.

**Comment by nzakas** [comment]
I wouldn't call this blindly copying, I think we've done a good job of poking at what's there and seeing what breaks. I'd rather start with something that has edge cases and start to fix them than keep waiting for a pristine rule with all the bells and whistles. This is the biggest gap between ESLint and both JSHint and JSCS, so I'd prefer to fill that gap even if it's not exactly correct yet.

On the other hand, if you (or anyone else) wants to start from scratch, or look at another option (https://github.com/nodeca/eslint-plugin-nodeca/blob/master/lib/indent.js), I'm all for it.

I just really want an indent rule sooner rather than later, and I don't think it has to be perfect to be useful (JSCS's limitations prove that).

File: tests/lib/rules/indent.js (hunk 5 above)
Comment (valorkin) [inline tests/lib/rules/indent.js]: @lo1tuma usually it is not a "good" idea to add lines at top of fixture files, for any tests

File: lib/rules/indent.js (hunk 12 above)
Comment (valorkin) [inline lib/rules/indent.js]: done

**Comment by valorkin** [comment]
@nzakas it seems to be done, from my side
review please

**Comment by lo1tuma** [comment]
@valorkin There is still this strange `break`-indent-detection behavior mentioned in [this comment](https://github.com/eslint/eslint/pull/1698#discussion_r23482962).

@nzakas I’m working on different implementation, but it is not feature complete yet. I try to push my approach this week.

**Comment by valorkin** [comment]
@lo1tuma it is optional now

**Comment by lo1tuma** [comment]
@valorkin I can’t find the option. How is it called?

**Comment by valorkin** [comment]
@lo1tuma `indentSwitchCase`
https://github.com/eslint/eslint/pull/1698#discussion_r23720231

**Comment by lo1tuma** [comment]
I thought this option is about the indentation of a a `SwitchCase` and not about a `BreakStatement`.

**Comment by valorkin** [comment]
@lo1tuma we can document and implement it in next change, because it can be tricky
```js
switch(a) {
  case 1: // no push indent
  case 2: break; // !isMultiline - no push indent
  case 3:
    b(); break; // should be invalid or valid? (IMHO invalid)
  case 4:
    b();
    break; // valid
  case 5:
    return;
  case 6:
    throw;
  default: 
    console.log('ups');
}
```
ps. using this indent on my production projects, all good

**Comment by nzakas** [comment]
I'm not sure how big of a deal the `break` alignment is for the first go-around. I'd be happy splitting that off into a separate enhancement.

**Comment by valorkin** [comment]
@nzakas +1 :)

File: docs/rules/indent.md [hunk 14]
```diff
@@ -0,0 +1,82 @@
+# This option sets a specific tab width for your code (indent)
```
Comment (nzakas) [inline docs/rules/indent.md]: Validate Indentation (indent)

File: docs/rules/indent.md [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+
+This option validates a specific tab width for your code.
```
Comment (nzakas) [inline docs/rules/indent.md]: This section should talk a bit about indentation styles.

File: docs/rules/indent.md [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+
+## Rule Details
+
+This rule is aimed at increasing code readability by enforcing consistent indentation style. The default style is `4 spaces`.
```
Comment (nzakas) [inline docs/rules/indent.md]: Let's remove "readability" from this. It's really just about ensuring consistent indentation.

File: docs/rules/indent.md [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+    b=c;
+    function(d) {
+        e=f;
+    }
+}
+```
+
+### Options
+
+The `indent` rule has two optional options:
```
Comment (nzakas) [inline docs/rules/indent.md]: two options

File: docs/rules/indent.md [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+* Indentation style, positive `Number` or `tab` (see rule detail for samples)
```
Comment (nzakas) [inline docs/rules/indent.md]: No need to use `Number`, just say number.

File: docs/rules/indent.md [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+* Configuring optional validations, `Object`.
+    * `indentSwitchCase` - validation switch case statements, `false` by default.
```
Comment (nzakas) [inline docs/rules/indent.md]: This actually means "indent switch cases".

**Comment by nzakas** [comment]
Okay, I think if we just clean up the documentation, we should be good to go.

@lo1tuma when you have a PR ready, we can consider it as well.

**Comment by valorkin** [comment]
@nzakas is it ok now?

File: docs/rules/indent.md [hunk 20]
```diff
@@ -0,0 +1,82 @@
+# Validate Indentation (indent)
+
+This option validates a specific tab width for your code in block statements.
```
Comment (nzakas) [inline docs/rules/indent.md]: This section should talk about indentation and indentation styles. You should why indentation is used and what the various styles are.

File: docs/rules/indent.md [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+
+```js
+// 4 spaces indention
+"indent": 2
+
+// 2 spaces indentation
+"indent": [2, 2]
+
+// tabs indentation
+"indent": [2, "tabs"]
```
Comment (nzakas) [inline docs/rules/indent.md]: Shouldn't this be "tab"?

File: docs/rules/indent.md [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+
+```js
+// 2 spaces indentation with enabled switch cases validation
+ "indent": [2, 2, {indentSwitchCase: true}]
+```
+
+## Compatibility
+
+* **JSHint**: `indent`
+* **jscs**: `validateIndentation`
```
Comment (nzakas) [inline docs/rules/indent.md]: JSCS (capitalized).

**Comment by nzakas** [comment]
Still need more in the intro section of the docs. Look at some examples on our site http://eslint.org/docs/rules/. 

Also, try turning this on for ESLint (top-level `.eslintrc` file) with a four-space indentation. If the rules is good, we shouldn't have that many warnings.

**Comment by valorkin** [comment]
@nzakas interesting that this rule checks more than `jshint`
```js
    if (lastReturn.code !== 0) {
      errors++;
    }
```
it is obvious that here we have inconsistent indentation,
I can make one more PR with fixes, to eslint code base
to have consistent indentation

more of that it would be good to check validation of 
- variable declarations with coma
- object declarations
in jscs it separate rules by the way

File: docs/rules/indent.md (hunk 20 above)
Comment (valorkin) [inline docs/rules/indent.md]: @nzakas `why indentation is used' - to make code readable?
added various styles mentioning

**Comment by valorkin** [comment]
@nzakas "fixing indentation issues"? docs updated, please check

**Comment by doberkofler** [comment]
Can't wait to use this rule and finally get rid of JSHint and jscs. Thank you @valorkin!

**Comment by gyandeeps** [comment]
Eagerly waiting for this rule to kick in... thanks @valorkin 
:+1: 
:100:

**Comment by nzakas** [comment]
Okay, this is in good enough shape to merge. @valorkin please start another PR to fix indentation issues in the ESLint project. We tend to find a lot of bugs when we turn rules on for the first time, so it would be good to do that before we release. Thanks!

**Comment by valorkin** [comment]
@nzakas it will be done

**Comment by mikesherov** [comment]
> I just really want an indent rule sooner rather than later, and I don't think it has to be perfect to be useful (JSCS's limitations prove that).

FYI, the edge cases that JSCS miss are purposeful. This rule is intended to not be a monotholic indentation rule, but rather a base engine that eventually other JSCS indentation rules will leverage. 

Missing multiline call expressions, etc. were in old versions of the rule and everyone complained because no one does it in the same way. The permissiveness of this rule is a design decision.

**Comment by mikesherov** [comment]
> We tend to find a lot of bugs when we turn rules on for the first time, so it would be good to do that before we release. Thanks!

Maybe you can send patches upstream to JSCS when you find them!

**Comment by puzrin** [comment]
Glad to see this rule in upstream. Will be waiting for 0.14 to switch from plugin.

**Comment by nzakas** [comment]
To address concerns from the JSCS team, I decided to revert this commit. We'll need to come up with a different implementation for ESLint.

**Comment by nzakas** [comment]
Per @mikesherov's comment on #1774, he's given his permission to use the code. I'm going to submit a new PR with this code and the updated attribution.
